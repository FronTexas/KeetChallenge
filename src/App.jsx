import React, { Component } from 'react';
import { ApolloClient, ApolloProvider, createNetworkInterface } from 'react-apollo';

import logo from './logo.svg';
import './App.css';

const networkInterface = createNetworkInterface({
  uri: 'http://localhost:8080',
});

const client = new ApolloClient({
  networkInterface,
  connectToDevTools: true,
});

class App extends Component {
  render() {
    return (
      <ApolloProvider client={client}>
        <div className="App">
          <div className="App-header">
            <img src={logo} className="App-logo" alt="logo" />
            <h2>Welcome to React</h2>
          </div>
          <p className="App-intro">
            To get started, edit <code>src/App.js</code> and save to reload.
          </p>
        </div>
      </ApolloProvider>
    );
  }
}

export default App;
