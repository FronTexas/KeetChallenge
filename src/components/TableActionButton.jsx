import React from 'react';
import { graphql } from 'react-apollo';
import PropTypes from 'prop-types';
import { Button } from 'react-md';


const TableActionButton = ({ mobile, children, mutate, OnClickHandler, ...props }) => {
  return (
    <Button
      icon={mobile}
      flat={!mobile}
      primary
      tooltipLabel={mobile ? children : null}
      onClick={() => OnClickHandler(mutate)}
      {...props}
    >
      {mobile ? null : children}
    </Button>
  );
};

TableActionButton.propTypes = {
  mobile: PropTypes.bool,
  children: PropTypes.node.isRequired,
  OnClickHandler: PropTypes.func.isRequired,
  mutate: PropTypes.func.isRequired,
};

TableActionButton.defaultProps = { mobile: false };

export default mutation => graphql(mutation)(TableActionButton);

