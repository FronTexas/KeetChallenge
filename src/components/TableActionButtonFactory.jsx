import React from 'react';
import { graphql } from 'react-apollo';
import PropTypes from 'prop-types';
import { Button } from 'react-md';


const TableActionButton = ({ mobile, children, mutate, OnClickHandler, ...props }) => (
  <Button
    icon={mobile}
    flat={!mobile}
    primary
    tooltipLabel={mobile ? children : null}
    onClick={() => OnClickHandler(mutate)}
    {...props}
  >
    {mobile ? null : children}
  </Button>
);

TableActionButton.defaultProps = { mobile: false };

TableActionButton.propTypes = {
  mobile: PropTypes.bool,
  children: PropTypes.node.isRequired,
  OnClickHandler: PropTypes.func.isRequired,
  mutate: PropTypes.func.isRequired,
};

export default mutation => graphql(mutation)(TableActionButton);
