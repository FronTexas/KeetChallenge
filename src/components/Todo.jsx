import React, { Component } from 'react';
import {
  TableRow,
  EditDialogColumn,
  Button,
} from 'react-md';
import { gql, graphql, compose } from 'react-apollo';


import PropTypes from 'prop-types';

const MUTATE_TITLE_TIMEOUT_INTERVAL = 300;

const mutateTitleMutationString = gql`
mutation mutateTitle($id: String!,$title: String!){
  save(id: $id, title: $title){
    id,title,completed
  }
}`;

const toggleCompletedMutationString = gql`
mutation toggleCompleted($id:String!){
  toggle(id:$id){
    id,title,completed
  }
}`;

const deleteTodoMutationString = gql`
mutation deleteTodo($id:String!){
  destroy(id:$id){
    id,title,completed
  }
}`;


class Todo extends Component {
  constructor(props) {
    super(props);
    this.mutateTitle = this.mutateTitle.bind(this);
    this.toggleCompleted = this.toggleCompleted.bind(this);
    this.deleteTodo = this.deleteTodo.bind(this);
  }

  mutateTitle(id, title, mutation) {
    if (this.timeOutId) clearTimeout(this.timeOutId);
    this.timeOutId = setTimeout(() => {
      mutation({ variables: { id, title } });
    }, MUTATE_TITLE_TIMEOUT_INTERVAL);
  }

  toggleCompleted(id, mutation) {
    mutation({ variables: { id } });
  }

  deleteTodo(id, refetch, mutation) {
    mutation({ variables: { id } })
      .then((res) => {
        if (res.data.destroy) {
          refetch();
        }
      });
  }

  render() {
    const { id, title, completed, refetch, mutateTitleMutation, toggleCompletedMutation, deleteTodoMutation } = this.props;
    return (
      <TableRow
        onCheckboxClick={() => { this.toggleCompleted(id, toggleCompletedMutation); }}
        selected={completed}
      >
        <EditDialogColumn
          defaultValue={title}
          placeholder="Type here to add new todo list"
          onChange={(updatedTitle) => { this.mutateTitle(id, updatedTitle, mutateTitleMutation); }}
          inline
        />
        <Button
          onClick={() => this.deleteTodo(id, refetch, deleteTodoMutation)}
          raised
          secondary
          iconChildren="close"
        >Delete</Button>
      </TableRow>
    );
  }
}

Todo.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
  refetch: PropTypes.func.isRequired,
  mutateTitleMutation: PropTypes.func.isRequired,
  toggleCompletedMutation: PropTypes.func.isRequired,
  deleteTodoMutation: PropTypes.func.isRequired,
};

export default compose(
  graphql(mutateTitleMutationString, { name: 'mutateTitleMutation' }),
  graphql(toggleCompletedMutationString, { name: 'toggleCompletedMutation' }),
  graphql(deleteTodoMutationString, { name: 'deleteTodoMutation' }),

)(Todo);
