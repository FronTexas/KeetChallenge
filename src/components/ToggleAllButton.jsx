import React from 'react';
import { gql, graphql } from 'react-apollo';
import PropTypes from 'prop-types';
import { Button } from 'react-md';


const toggleAllMutation = gql`
mutation toggleAll($checked:Boolean!){
  toggleAll(checked:$checked){
    id,title,completed
  }
}`;

const ToggleAllButton = ({ mobile, children, mutate, onToggleAllClick, ...props }) => {
  return (
    <Button
      icon={mobile}
      flat={!mobile}
      primary
      tooltipLabel={mobile ? children : null}
      onClick={() => onToggleAllClick(mutate)}
      {...props}
    >
      {mobile ? null : children}
    </Button>
  );
};

ToggleAllButton.propTypes = {
  mobile: PropTypes.bool,
  children: PropTypes.node.isRequired,
  onToggleAllClick: PropTypes.func.isRequired,
  mutate: PropTypes.func.isRequired,
};

ToggleAllButton.defaultProps = { mobile: false };

export default graphql(toggleAllMutation)(ToggleAllButton);

