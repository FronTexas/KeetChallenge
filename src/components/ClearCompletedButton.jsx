import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import { CLEAR_COMPLETED } from '../graph';

import TableActionButton from './TableActionButton';

export class ClearCompletedButtonComponent extends Component {
  static propTypes = {
    clearCompleted: PropTypes.func.isRequired,
    refetchTodos: PropTypes.func.isRequired,
  }

  handleOnClick = () => {
    const { clearCompleted, refetchTodos } = this.props;
    clearCompleted()
      .then((res) => {
        if (res.data.clearCompleted) {
          refetchTodos();
        }
      });
  }

  render() {
    return (
      <TableActionButton onClick={this.handleOnClick} {...this.props} />
    );
  }
}

export default graphql(CLEAR_COMPLETED, {
  props: ({ mutate }) => ({
    clearCompleted: () => mutate(),
  }),
})(ClearCompletedButtonComponent);

