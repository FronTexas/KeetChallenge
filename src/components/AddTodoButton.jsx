import { gql } from 'react-apollo';

import TableActionButtonFactory from './TableActionButton';

const AddTodoButton = TableActionButtonFactory(gql`
mutation addTodo($title:String!){
	add(title:$title){
	  id,title,completed
	}
}`);

export default AddTodoButton;
