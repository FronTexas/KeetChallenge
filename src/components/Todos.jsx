import React, { Component } from 'react';
import { gql, graphql } from 'react-apollo';

import {
  DataTable,
  TableBody,
  TableCardHeader,
} from 'react-md';

import Todo from './Todo';
import AddTodoButton from './AddTodoButton';
import ToggleAllButton from './ToggleAllButton';
import ClearCompletedButton from './ClearCompletedButton';


const queryTodos = gql`{
  todos{
    id
    title
    completed
  }
}`;

class Todos extends Component {
  constructor(props) {
    super(props);
    this.handleAddTodoClick = this.handleAddTodoClick.bind(this);
    this.handleToggleAllClick = this.handleToggleAllClick.bind(this);
    this.handleClearCompletedClick = this.handleClearCompletedClick.bind(this);
    this.state = {
      toggleAllState: true,
    };
  }

  handleAddTodoClick(mutation, refetch) { 
    mutation({ variables: { title: '' } })
      .then((res) => {
        if (res.data.add) {
          refetch();
        }
      });
  }

  handleToggleAllClick(mutation) {
    mutation({ variables: { checked: this.state.toggleAllState } });
    this.setState({ toggleAllState: !this.state.toggleAllState });
  }

  handleClearCompletedClick(mutation, refetch) {
    mutation()
      .then((res) => {
        if (res.data.clearCompleted) {
          refetch();
        }
      });
  }

  render() {
    const { data: { loading, error, refetch, todos } } = this.props;
    if (loading) { return <p>Loading...</p>; }
    if (error) { return <p>Error!</p>; }
    return (
      <div>
        <DataTable baseId="todos" fullWidth={false} className="todos">
          <TableCardHeader
            title="To-do list"
          >
            <AddTodoButton onAddTodoClick={mutation => this.handleAddTodoClick(mutation, refetch)} iconChildren="add" >Add</AddTodoButton>
            <ToggleAllButton onToggleAllClick={this.handleToggleAllClick}  toggleAllState={false} iconChildren="done">Toggle All</ToggleAllButton>
            <ClearCompletedButton onClearCompletedClick={mutation => this.handleClearCompletedClick(mutation, refetch)} iconChildren="clear">Clear Completed</ClearCompletedButton>
          </TableCardHeader>
          <TableBody>
            {
              todos.map((item) => {
                const { id, title, completed } = item;
                return <Todo refetch={refetch} key={id} id={id} title={title} completed={completed} />;
              })
            }
          </TableBody>
        </DataTable>
      </div>
    );
  }
}

export default graphql(queryTodos)(Todos);
