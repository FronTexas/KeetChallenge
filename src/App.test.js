import React from 'react';
import { shallow, mount } from 'enzyme';
import { gql } from 'react-apollo';
import { Todos } from './components/Todos';
import App from './App';
import { Todo } from './components/Todo';
import TableActionButton from './components/TableActionButton';
import { AddTodoButton } from './components/AddTodoButton';
import { ClearCompletedButton } from './components/ClearCompletedButton';
import { ToggleAllButton } from './components/ToggleAllButton';

it('shallow renders without crashing', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.length).toEqual(1);
});

// describe('Test for <Todos/>', () => {
//   let wrapper;
//   const mockedFunctions = {
//     handleAddTodoClick: jest.fn(),
//     handleToggleAllClick: jest.fn(),
//     handleClearCompletedClick: jest.fn(),
//     handleDeleteTodo: jest.fn(),
//     handleToggleCompleted: jest.fn(),
//     handleMutateTitle: jest.fn(),
//   };

//   beforeEach(() => {
//     const mockedData = {
//       loading: false,
//       error: false,
//       refetch: () => {},
//       todos: [{ id: '123', title: 'Empty', completed: false }],
//     };
//     wrapper = mount(<Todos data={mockedData} />);
//     wrapper.instance().handleAddTodoClick = mockedFunctions.handleAddTodoClick;
//     wrapper.instance().handleToggleAllClick = mockedFunctions.handleToggleAllClick;
//     wrapper.instance().handleClearCompletedClick = mockedFunctions.handleClearCompletedClick;
//     wrapper.instance().handleDeleteTodo = mockedFunctions.handleDeleteTodo;
//     wrapper.instance().handleToggleCompleted = mockedFunctions.handleToggleCompleted;
//     wrapper.instance().handleMutateTitle = mockedFunctions.handleMutateTitle;
//   });

//   it('Calls AddTodo handler', () => {
//     wrapper.find('#add-todo-button').props().onClick();
//     expect(mockedFunctions.handleAddTodoClick).toHaveBeenCalled();
//   });

//   it('Calls ClearCompleted handler', () => {
//     wrapper.find('#clear-completed-button').props().onClick();
//     expect(mockedFunctions.handleClearCompletedClick).toHaveBeenCalled();
//   });

//   it('Calls MutateTitle handler', () => {
//     wrapper.find('Todo').props().mutateTitle();
//     expect(mockedFunctions.handleMutateTitle).not.toHaveBeenCalled();
//   });
// });

describe('Test for Todo', () => {
  let wrapper;
  const minProps = {
    id: '',
    title: '',
    completed: false,
    refetchTodos: () => {},
    toggleTodo: jest.fn(),
    mutateTitle: jest.fn(),
    deleteTodo: jest.fn(),
  };

  beforeEach(() => {
    wrapper = mount(<Todo {...minProps} />);
  });

  it('Renders Todo without exploding', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('Calls deleteTodo when delete button is pressed', () => {
    const mockPromise = new Promise((resolve) => {
      resolve();
    });
    minProps.deleteTodo.mockReturnValueOnce(mockPromise);
    wrapper.find('Button').simulate('click');
    expect(minProps.deleteTodo).toHaveBeenCalled();
  });

  it('Calls toggleTodo when the checkbox is clicked', () => {
    wrapper.find('TableRow').props().onCheckboxClick();
    expect(minProps.toggleTodo).toHaveBeenCalled();
  });

  it('Calls mutateTitle after editing the title', () => {
    jest.useFakeTimers();
    wrapper.find('EditDialogColumn').props().onChange();
    jest.runAllTimers();
    expect(minProps.mutateTitle).toHaveBeenCalled();
  });
});

describe('Test for TableActionButton', () => {
  let wrapper;
  const minProps = {
    mobile: false,
  };

  beforeEach(() => {
    wrapper = mount(<TableActionButton{...minProps}>Test</TableActionButton>);
  });

  it('Renders TableActionButton without exploding', () => {
    expect(wrapper.length).toEqual(1);
  });
});

describe('Test for AddTodoButton', () => {
  let wrapper;
  let addTodo;
  let mockPromise;
  let refetchTodos;

  beforeEach(() => {
    addTodo = jest.fn();
    mockPromise = new Promise((resolve) => {
      const res = { data: { add: true } };
      resolve(res);
    });
    refetchTodos = jest.fn();
    addTodo.mockReturnValueOnce(mockPromise);
    wrapper = mount(<AddTodoButton addTodo={addTodo} refetchTodos={refetchTodos} />);
  });

  it('Calls addTodo when clicked', () => {
    wrapper.find('TableActionButton').props().onClick();
    expect(addTodo).toHaveBeenCalled();
  });

  it('Calls refetchTodos after calling addTodo', () => {
    wrapper.find('TableActionButton').props().onClick();
    expect(refetchTodos).toHaveBeenCalled();
  });
});

describe('Test for ClearCompletedButton', () => {
  let wrapper;
  let clearCompleted;
  let mockPromise;
  let refetchTodos;

  beforeEach(() => {
    clearCompleted = jest.fn();
    mockPromise = new Promise((resolve) => {
      const res = { data: { clearCompleted: true } };
      resolve(res);
    });
    refetchTodos = jest.fn();
    clearCompleted.mockReturnValueOnce(mockPromise);
    wrapper = mount(<ClearCompletedButton clearCompleted={clearCompleted} refetchTodos={refetchTodos} />);
  });

  it('Calls clearCompleted when clicked', () => {
    wrapper.find('TableActionButton').props().onClick();
    expect(clearCompleted).toHaveBeenCalled();
  });

  it('Calls refetchTodos after calling addTodo', () => {
    wrapper.find('TableActionButton').props().onClick();
    expect(refetchTodos).toHaveBeenCalled();
  });
});

describe('Test for ToggleAllButton', () => {
  let wrapper;
  let toggleAllTodos;
  let mockPromise;

  beforeEach(() => {
    toggleAllTodos = jest.fn();
    mockPromise = new Promise((resolve) => {
      resolve();
    });
    toggleAllTodos.mockReturnValueOnce(mockPromise);
    wrapper = mount(<ToggleAllButton toggleAllTodos={toggleAllTodos} />);
  });

  it('Calls toggleAllTodos when clicked', () => {
    wrapper.find('TableActionButton').props().onClick();
    expect(toggleAllTodos).toHaveBeenCalled();
  });
});
