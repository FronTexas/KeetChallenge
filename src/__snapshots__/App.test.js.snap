// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow renders without crashing 1`] = `
<ApolloProvider
  client={
    ApolloClient {
      "addTypename": true,
      "dataId": [Function],
      "dataIdFromObject": [Function],
      "disableNetworkFetches": false,
      "fieldWithArgs": [Function],
      "fragmentMatcher": HeuristicFragmentMatcher {},
      "initialState": Object {},
      "middleware": [Function],
      "mutate": [Function],
      "networkInterface": HTTPFetchNetworkInterface {
        "_afterwares": Array [],
        "_middlewares": Array [],
        "_opts": Object {},
        "_uri": "http://localhost:8080",
      },
      "query": [Function],
      "queryDeduplication": true,
      "reducerConfig": Object {
        "addTypename": true,
        "customResolvers": undefined,
        "dataIdFromObject": [Function],
        "fragmentMatcher": [Function],
      },
      "resetStore": [Function],
      "setStore": [Function],
      "ssrMode": false,
      "version": "1.9.2",
      "watchQuery": [Function],
    }
  }
>
  <div
    className="App"
  >
    <div
      className="App-header"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h2>
        Welcome to React
      </h2>
    </div>
    <p
      className="App-intro"
    >
      To get started, edit 
      <code>
        src/App.js
      </code>
       and save to reload.
    </p>
  </div>
</ApolloProvider>
`;
